tests/java/verifiers/src/main/java/io/vectorized/compaction/idempotency/Workload.java

```diff
@@ -0,0 +1,522 @@
+package io.vectorized.compaction.idempotency;
+import java.lang.Thread;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.Semaphore;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.producer.Callback;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+import org.apache.kafka.common.TopicPartition;
+
+public class Workload {
+  static enum OpStatus { WRITING, UNKNOWN, WRITTEN }
+
+  static enum State { INITIAL, PRODUCING, TRANSITIONAL, CONSUMING, FINAL }
+
+  static class WriteInfo {
+    public long id;
+    public String key;
+    public int partition;
+
+    public OpStatus status;
+    public long offset;
+  }
+
+  static class LatestValue {
+    public WriteInfo confirmed;
+    public Queue<WriteInfo> attempts;
+
+    public LatestValue() {
+      this.confirmed = null;
+      this.attempts = new LinkedList<>();
+    }
+  }
+
+  static class Value {
+    public long id;
+    public long offset;
+
+    public Value(long id, long offset) {
+      this.id = id;
+      this.offset = offset;
+    }
+  }
+
+  static class Partition {
+    public int id;
+    public Queue<Long> writeLog;
+    public Semaphore feedback;
+    public HashMap<Long, WriteInfo> writes;       // hashed by WriteInfo.id
+    public HashMap<String, LatestValue> latestKV; // hashed by key
+
+    public boolean draining = false;
+    public boolean hasReader = false;
+    public int inflights = 0;
+    public long countWritten = 0;
+    public long countRead = 0;
+    public Semaphore drained = new Semaphore(0);
+
+    public Partition(int id) {
+      this.id = id;
+      this.writeLog = new LinkedList<>();
+      this.feedback = new Semaphore(200);
+      this.writes = new HashMap<>();
+      this.latestKV = new HashMap<>();
+    }
+  }
+
+  private static final String LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+  public volatile State state = State.INITIAL;
+  HashMap<Integer, Partition> partitions;
+  private volatile App.InitBody args;
+  volatile long started_s;
+  public volatile boolean are_writers_stopped = false;
+  private volatile ArrayList<Thread> write_threads;
+  private volatile ArrayList<Thread> read_threads;
+  private long count = 0;
```

Comment:
i think the only variable here that needs to be `volatile` is `are_writers_stopped`

Reply:
and `state`

Reply:
`waitProducer` and `startProducer` both access `write_threads` & `read_threads` but there is no explicit happens-before relation between them

Reply:
`volatile` in those case doesn't change anything. Marking an `ArrayList` volatile doesn't make all the updates to it immediately visible to other other threads. Only an update of reference itself is forcing visibility of all the actions that happened in the same thread before the update. If there is no relation between `waitProducer` and `startProducer` we should use a synchronization primitive.